### PR TITLE
chore(regression): regression test old gdk version

### PR DESCRIPTION
**Pull request by devnaruka**
**Issue #, if available:**

**Description of changes:**
Add multiple gdk versions to verify regression between latest and old gdk versions.

1. Add new option --target-version to pass the expected the gdk-cli versions to UATs.
   * Example:
      *  `pytest -v -s uat --target-version 1.0.0`
      *  `pytest -v -s uat --target-version 1.1.0`
      *  `pytest -v -s uat --target-version HEAD`
   * If not provided then target version defaults to `HEAD`, which translates to the code at the most recent commit against which the regression workflow is running.
2. Add `@pytest.mark.version(...)` test marker to provide version constraints for any UAT test.
   * The marker takes `kwargs` (key & values) to define constraints. These can be combined to create upper and lower bounds. Available key args are:
      * `eq`: equal to gdk cli version (exact version match).
      * `min` : minimum gdk cli version (lower bound)
      * `ge`: greater than or equal to gdk cli version (lower bound). Alias of `min`.
      * `gt`: greater than gdk cli version (lower bound).
      * `max` : maximum gdk cli version (upper bound)
      * `le`: less than or equal to gdk cli version (upper bound). Alias of `max`.
      * `lt`: less than gdk cli version (upper bound).
   * Examples:
       * `@pytest.mark.version(eq='1.1.0')`
       * `@pytest.mark.version(min='1.1.0')` or `@pytest.mark.version(ge='1.1.0')`
       * `@pytest.mark.version(max='1.1.0')` or `@pytest.mark.version(le='1.1.0')`
       * `@pytest.mark.version(gt='1.1.0')`
       * `@pytest.mark.version(lt='1.1.0')`
       * `@pytest.mark.version(min='1.0.0', max='1.1.0')`
 3. Applied version constraints based on feature changes between versions.

**Why is this change necessary:**
To ensure new changes are not breaking older versions.

**How was this change tested:**
With workflows runs as part of this pull request.

**Any additional information or context required to review the change:**

**Checklist:**
- [ ] Updated the README if applicable
- [ ] Updated or added new unit tests
- [ ] Updated or added new integration tests
- [x] Updated or added new end-to-end tests
- [ ] If your code makes a remote network call, it was tested with a proxy
- [x] You confirm that the change is backwards compatible

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: uat/test_uat_build.py [hunk 1]
```diff
@@ -35,10 +39,40 @@ def test_build_template_zip(change_test_dir, gdk_cli):
         .joinpath("HelloWorld.zip")
         .resolve()
     )
-
     assert artifact_path.exists()
 
 
+def test_build_template_zip_fail_with_no_artifact_since_v1_0_0(change_test_dir, gdk_cli):
```
Comment (saranyailla) [inline uat/test_uat_build.py:45]: instead of adding `since_` here, can we mark it as `min='1.0.0'`?

Comment (devnaruka) [reply uat/test_uat_build.py:45]: done

File: uat/conftest.py [hunk 2]
```diff
@@ -21,8 +25,61 @@ def change_test_dir(tmpdir, monkeypatch):
...
+    # request.node is the current test item
+    # query the marker "version" of current test item
+    version_marker = request.node.get_closest_marker('version')
+
+    # if test item was not marked, there's no version restriction
+    if version_marker is None:
+        return
+
+    # parse target version from options
+    target_version = version_parser(version_option) if not latest_mode else version_option
```
Comment (saranyailla) [inline uat/conftest.py:53]: Why would target_version be either of `version_parser(version_option)` and `version_option`?

File: uat/conftest.py [hunk 3]
```diff
@@ -21,8 +25,61 @@ def change_test_dir(tmpdir, monkeypatch):
...
+
+    # if used as @pytest.mark.version(eq=1.0.0)
+    eq_version = version_marker.kwargs.get('eq')
+    eq_version = version_parser(eq_version) if eq_version is not None else None
```
Comment (saranyailla) [inline uat/conftest.py:57]: if `version_parser(None) `is None, then can L56 and 57 be combined?

File: uat/conftest.py [hunk 4]
```diff
@@ -21,8 +25,61 @@ def change_test_dir(tmpdir, monkeypatch):
...
+    if eq_version is not None and (latest_mode or target_version == eq_version):
```
Comment (saranyailla) [inline uat/conftest.py:58]: Here and everywhere, `if x is not None` can be shortened as `if x`. 
Also, if there's a possibility of adding `version(eq='')`accidentally on tests which should be treated as None, `if x` is the way to go unless `version_parser` has a diff behavior with `None` vs `''`.

File: uat/conftest.py (hunk 2 above)
Comment (devnaruka) [reply uat/conftest.py:53]: This simplifies the messages passed to pytest.skip(...) later during the version constraints evaluates.

if its `latest_mode` then `target_version` would evaluate to 'HEAD' string else a `Version` object.

File: uat/conftest.py (hunk 3 above)
Comment (devnaruka) [reply uat/conftest.py:57]: I check for None beforehand as `version_parser(None)` throws error.

**Review by pallavi03 — APPROVED** [review]

File: uat/conftest.py [hunk 5]
```diff
@@ -21,8 +25,61 @@ def change_test_dir(tmpdir, monkeypatch):
...
+    version_option = request.config.getoption('--target-version')
+    if version_option is None:
+        return
+
+    latest_mode = version_option == 'HEAD'
+    print(f"\nTarget gdk-cli version {version_option}")
+
+    # request.node is the current test item
+    # query the marker "version" of current test item
+    version_marker = request.node.get_closest_marker('version')
```
[the remark this answers is not in the document]
Comment (devnaruka) [reply uat/conftest.py:46]: If both upper and lower bounds provided then, both would be evaluated and enforced.

**Review by saranyailla — DISMISSED** [review]

File: uat/conftest.py (hunk 4 above)
Comment (devnaruka) [reply uat/conftest.py:58]: I like the shortened version. I will update.

File: uat/README.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+```shell
+python3 -m pip install pip
+pip install .
+```
+
+#### 3. Configure AWS credentials
+
+Configure AWS CLI with your credentials as shown here - https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html.
+
+#### 4. Installing test dependencies
```
Comment (saranyailla) [inline uat/README.md:34]: IMO, this can be moved to a separate README inside github workflows or uat folder. This is the landing page for customers and it has more information than needed to install and work with gdk. :)

File: uat/README.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
+# End-to-end tests for Greengrass Development Kit (GDK) - Command Line Interface (CLI)
+
```
Comment (saranyailla) [inline uat/README.md:2]: Thanks for adding this README!!

**Review by saranyailla — APPROVED** [review]

**Review by pallavi03 — APPROVED** [review]
